classes/task/sendfailednotifications_task.php

```diff
@@ -0,0 +1,168 @@
+
+        $sql = "SELECT x.id, x.errortype AS type, x.eventname, u.firstname, u.lastname, x.contextid, x.response, x.timecreated
+        FROM {logstore_xapi_failed_log} x
+        LEFT JOIN {user} u
+        ON u.id = x.userid";
+
+        $results = $DB->get_records_sql($sql);
+        return $results;
+    }
+
+    /**
+     * Get failed email message.
+     *
+     * @return string email message in html
+     */
+    private function get_failed_email_message($results) {
+        $emailmsg = "";
+
+        $emailmsg .= "<p>".PHP_EOL;
+        $emailmsg .= get_string('failedtosend', 'logstore_xapi').PHP_EOL;
+        $emailmsg .= "</p>".PHP_EOL;
+
+        // summary info
+        $emailmsg .= "<table>".PHP_EOL;
+
+        $endpointname = get_string('endpoint', 'logstore_xapi');
+        $url = get_config('logstore_xapi', 'endpoint');
+        $endpointurl = "<a href='$url' target='_blank'>$url</a>";
+
+        $errorlogpage = get_string('errorlogpage', 'logstore_xapi');
+        $url = new \moodle_url("/admin/tool/log/store/xapi/report.php");
+        $errorlogurl = "<a href='$url' target='_blank'>$url</a>";
+
+        $row = "<tr>";
+        $row .= "<td>".$endpointname."</td>";
+        $row .= "<td>".$endpointurl."</td>";
+        $row .= "</tr>".PHP_EOL;
+        $emailmsg .= $row;
+
+        $row = "<tr>";
+        $row .= "<td>".$errorlogpage."</td>";
+        $row .= "<td>".$errorlogurl."</td>";
+        $row .= "</tr>".PHP_EOL;
+        $emailmsg .= $row;
+
+        $emailmsg .= "</table>".PHP_EOL;
+
+        // separator
+        $emailmsg .= "<h3>".get_string('failurelog', 'logstore_xapi')."</h3>".PHP_EOL;
+
+        // rows
+        $emailmsg .= "<table>".PHP_EOL;
+
+        $emailmsg .= '<tr>'.PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('datetimegmt', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('eventname', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('response', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "</tr>".PHP_EOL;
+
+        foreach ($results as $result) {
+            $row = "<tr>";
+            $row .= "<td>".userdate($result->timecreated)."</td>";
+            $row .= "<td>".$result->eventname."</td>";
+            $row .= "<td>".$result->response."</td>";
+            $row .= "</tr>".PHP_EOL;
+
+            $emailmsg .= $row;
+        }
+        $emailmsg .= "</table>".PHP_EOL;
+
+        return $emailmsg;
+    }
+
+    /**
+     * Send email using email_to_user.
+     *
+     * @return int 1 = sent, 0 = not sent
+     */
+    private function sendmail($msg, $subject, $emailto) {
+ 
+        $user = new \stdClass();
+        $user->id = 1;
+        $user->username = "";
+        $user->email = $emailto;
+        $user->deleted = 0;
+        $user->mailformat = FORMAT_HTML;
+
+        $from = new \stdClass();
+        $from->id = 1;
+        $from->username = "";
+        $from->email = "";
```

Comment:
what will the sender information look like on the generated email if this is empty?

Reply:
It uses the global system set from address.